llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp

```diff
@@ -186,6 +186,22 @@ static Value *foldSelectICmpAnd(SelectInst &Sel, ICmpInst *Cmp,
       return ExtraBitInTC ? Builder.CreateXor(V, C) : Builder.CreateOr(V, C);
     }
 
+    // (V & AndMaskC) == 0 ? TC : FC --> (V & AndMaskC) + TC
+    if (TC + AndMask == FC) {
+      if (CreateAnd)
+        V = Builder.CreateAnd(V, ConstantInt::get(SelType, AndMask));
+      Constant *C = ConstantInt::get(SelType, TC);
+      return Builder.CreateAdd(V, C);
+    }
+
+    // (V & AndMaskC) == 0 ? TC : FC --> TC - (V & AndMaskC)
+    if (TC - AndMask == FC) {
+      if (CreateAnd)
+        V = Builder.CreateAnd(V, ConstantInt::get(SelType, AndMask));
+      Constant *C = ConstantInt::get(SelType, TC);
+      return Builder.CreateSub(C, V);
+    }
```

Comment:
IMO this can easily be generalized to all binops where 0 rhs is identity.

Reply:
TODO: Shl/LShr/AShr

Reply:
Think it would be easier to just loop through them and only write this code once i.e:

```
Constant * TCC = ConstantInt::get(TC);
Constant * FCC = ConstantInt::get(FC);
Constant * MaskC = ConstantInt::get(AndMask);

for(Opc : { Add, Sub, Shl...}) {
   if (SimplifyBinOp(Opc, TCC, MaskC) == FCC) {
      if (CreateAnd)
        V = Builder.CreateAnd(V, MaskC);
      return Builder.CreateBinOp(BinOp, V, C);
  }
}
```

Reply:
I know I have a tendency to over-generalize at the cost of complexity, but in this case I'd argue even two explicit cases is as much complexity as just a generalized loop.

Reply:
Before: http://llvm-compile-time-tracker.com/compare.php?from=fbd303ba5dd98d4513d62058bc1ad463507464ce&to=9f108f5a9cd6645212efdfff325eb484171799ab&stat=instructions%3Au

Reply:
Can you handle the or/xor cases in this loop too?